pkg/clustermesh/kvstoremesh/kvstoremesh_test.go

```diff
@@ -251,3 +256,114 @@ func TestRemoteClusterRun(t *testing.T) {
 		})
 	}
 }
+
+// mockClusterMesh is a mock implementation of the common.ClusterMesh interface
+// allowing for direct manipulation of the clusters
+type mockClusterMesh struct {
+	clusters map[string]*remoteCluster
+}
+
+// ForEachRemoteCluster is a mirrored implementation of ClusterMesh.ForEachRemoteCluster that operates on the mocked clusters.
+func (m *mockClusterMesh) ForEachRemoteCluster(fn func(common.RemoteCluster) error) error {
+	for _, cluster := range m.clusters {
+		if err := fn(cluster); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *mockClusterMesh) NumReadyClusters() int {
+	return len(m.clusters)
+}
+
+func (m *mockClusterMesh) Start(cell.HookContext) error {
+	return nil
+}
+
+func (m *mockClusterMesh) Stop(cell.HookContext) error {
+	return nil
+}
+
+func TestRemoteClusterSync(t *testing.T) {
+	tests := []struct {
+		name    string
+		connect bool
+		sync    bool
+	}{
+		{
+			name:    "remote cluster successfully syncs",
+			connect: true,
+			sync:    true,
+		},
+		{
+			name:    "remote cluster fails to connect",
+			connect: false,
+			sync:    false,
+		},
+		{
+			name:    "remote cluster connects but fails to sync",
+			connect: true,
+			sync:    false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+
+			mockClusterMesh := &mockClusterMesh{
+				clusters: make(map[string]*remoteCluster),
+			}
+			km := KVStoreMesh{
+				// Set a very low timeout to speed up the test
+				config: Config{PerClusterReadyTimeout: 1 * time.Millisecond, GlobalReadyTimeout: 1 * time.Millisecond},
+				common: mockClusterMesh,
+				logger: logrus.New(),
+			}
+
+			synced := newSynced()
+			synced.resources.Add()
+			synced.resources.Stop()
+
+			rc := &remoteCluster{
+				name:   "foo",
+				synced: synced,
+				logger: km.logger.WithField(logfields.ClusterName, "foo"),
+			}
+			mockClusterMesh.clusters[rc.name] = rc
+
+			if tt.connect {
+				close(rc.synced.connected)
+				synced.resources.Done()
+			}
+
+			// trigger the readiness timeout
+			rc.waitForConnection(ctx)
+
+			require.NoError(t, rc.synced.Resources(ctx), "Still waiting for remote cluster resources")
```

Comment:
This is intentional, since all resources should be completed by this point in both paths:
- The happy path is that the connection succeeds, making `rc.waitForConnection` a noop and returning immediately.
- The sad path is we don't establish a connection, and `rc.WaitForConnection` triggers the timeout, marking all resources as done.

There may be another angle I'm not seeing, though

Reply:
After re-reading your comment I realize I misunderstood. If the `tt.connect` branch is removed, cases where the conditional is true will still pass, but due to the timeout rather than actually being marked `Done()`. I'll update the test

Reply:
Updated. If the `tt.connect` conditional is removed, tests that expect `tt.connect` will fail